src/wasm/wasm-emscripten.cpp

```diff
@@ -207,6 +208,59 @@ void EmscriptenGlueGenerator::generateDynCallThunks() {
   }
 }
 
+static Function* ensureFunctionImport(Module* module, Name name, std::string sig) {
+  // Then see if its already imported
+  ImportInfo info(*module);
+  if (Function* f = info.getImportedFunction(ENV, name)) {
+    return f;
+  }
+  // Failing that create a new function import.
+  auto import = new Function;
+  import->name = name;
+  import->module = ENV;
+  import->base = name;
+  auto* functionType = ensureFunctionType(sig, module);
+  import->type = functionType->name;
+  FunctionTypeUtils::fillFunction(import, functionType);
+  module->addFunction(import);
+  return import;
+}
+
+struct RemoveStackPointer : public PostWalker<RemoveStackPointer> {
+  RemoveStackPointer(Global* StackPointer) : StackPointer(StackPointer) {}
+
+  void visitGetGlobal(GetGlobal* curr) {
+    if (getModule()->getGlobalOrNull(curr->name) == StackPointer) {
+      ensureFunctionImport(getModule(), STACK_SAVE, "i");
+      if (!builder) builder = make_unique<Builder>(*getModule());
+      replaceCurrent(builder->makeCall(STACK_SAVE, {}, i32));
+    }
+  }
+
+  void visitSetGlobal(SetGlobal* curr) {
+    if (getModule()->getGlobalOrNull(curr->name) == StackPointer) {
+      ensureFunctionImport(getModule(), STACK_RESTORE, "vi");
+      if (!builder) builder = make_unique<Builder>(*getModule());
+      replaceCurrent(builder->makeCall(STACK_RESTORE, {curr->value}, none));
+    }
+  }
+
+  std::unique_ptr<Builder> builder;
```

Comment:
Nit: These could be private. I think they should be because they're not part of the API, but it's really pretty moot.